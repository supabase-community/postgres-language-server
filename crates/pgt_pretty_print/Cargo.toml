[package]
authors.workspace    = true
categories.workspace = true
description          = "<DESCRIPTION>"
edition.workspace    = true
homepage.workspace   = true
keywords.workspace   = true
license.workspace    = true
name                 = "pgt_pretty_print"
repository.workspace = true
version              = "0.0.0"


[dependencies]
pgt_pretty_print_codegen.workspace = true
pgt_query.workspace                = true

[dev-dependencies]
