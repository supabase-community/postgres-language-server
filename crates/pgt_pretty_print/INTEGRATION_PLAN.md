# PostgreSQL Pretty Printer Integration Plan

## Current Status

The pretty printer foundation is **complete and working**! Basic SQL formatting is functional with:
- ✅ SELECT statements with aliases, schema qualification
- ✅ Line length-based breaking (configurable via filename suffix)
- ✅ Proper comma placement and indentation
- ✅ Comprehensive test suite with snapshot testing
- ✅ AST integrity verification (location-aware comparison)

## Architecture Overview

```
SQL Input → pgt_query::parse() → AST → ToTokens → Layout Events → Renderer → Formatted SQL
```

**Key Components:**
- **ToTokens trait**: Converts AST nodes to layout events
- **Layout Events**: `Token`, `Space`, `Line(Hard/Soft/SoftOrSpace)`, `GroupStart/End`, `IndentStart/End`
- **Renderer**: Two-phase prettier-style algorithm (try single line, else break)

## Renderer Implementation Status

### ✅ Completed
- **Core rendering pipeline**: Event processing, text/space/line output
- **Basic grouping**: Single-line vs multi-line decisions
- **Indentation**: Configurable spaces/tabs with proper nesting
- **Line length enforcement**: Respects `max_line_length` config
- **Token rendering**: Keywords, identifiers, punctuation
- **Break propagation**: Child groups with `break_parent: true` force parent groups to break
- **Nested group independence**: Inner groups make independent fit decisions when outer groups break
- **Stack overflow elimination**: Fixed infinite recursion in renderer

### ❌ Missing Features (Priority Order)

#### 1. **Group ID References** (Medium Priority)
**Issue**: Groups can't reference each other's break decisions.

```rust
// Missing: Conditional formatting based on other groups
GroupStart { id: Some("params") }
// ... later reference "params" group's break decision
```

**Implementation**:
- Track group break decisions by ID
- Add conditional breaking logic

#### 2. **Advanced Line Types** (Medium Priority)
**Issue**: `LineType::Soft` vs `LineType::SoftOrSpace` handling could be more sophisticated.

**Current behavior**:
- `Hard`: Always breaks
- `Soft`: Breaks if group breaks, disappears if inline
- `SoftOrSpace`: Breaks if group breaks, becomes space if inline

**Enhancement**: Better handling of soft line semantics in complex nesting.

#### 3. **Performance Optimizations** (Low Priority)
- **Early bailout**: Stop single-line calculation when length exceeds limit
- **Caching**: Memoize group fit calculations for repeated structures
- **String building**: More efficient string concatenation

## AST Node Coverage Status

### ✅ Implemented ToTokens
- `SelectStmt`: Basic SELECT with FROM clause
- `ResTarget`: Column targets with aliases
- `ColumnRef`: Column references (schema.table.column)
- `String`: String literals in column references  
- `RangeVar`: Table references with schema
- `FuncCall`: Function calls with break propagation support

### ❌ Missing ToTokens (Add as needed)
- `InsertStmt`, `UpdateStmt`, `DeleteStmt`: DML statements
- `WhereClause`, `JoinExpr`: WHERE conditions and JOINs
- `AExpr`: Binary/unary expressions (`a = b`, `a + b`)
- `AConst`: Literals (numbers, strings, booleans)
- `SubLink`: Subqueries
- `CaseExpr`: CASE expressions
- `WindowFunc`: Window functions
- `AggRef`: Aggregate functions
- `TypeCast`: Type casting (`::int`)

## Testing Infrastructure

### ✅ Current
- **dir-test integration**: Drop SQL files → automatic snapshot testing
- **Line length extraction**: `filename_80.sql` → `max_line_length: 80`
- **AST integrity verification**: Ensures no data loss during formatting
- **Location field handling**: Clears location differences for comparison

### 🔄 Enhancements Needed
- **Add more test cases**: Complex queries, edge cases
- **Performance benchmarks**: Large SQL file formatting speed
- **Configuration testing**: Different indent styles, line lengths
- **Break propagation testing**: Verified with `FuncCall` implementation

## Integration Steps

### ✅ Phase 1: Core Renderer Fixes (COMPLETED)
1. ✅ **Fix break propagation**: Implemented proper `break_parent` handling
2. ✅ **Fix nested groups**: Allow independent fit decisions  
3. ✅ **Fix stack overflow**: Eliminated infinite recursion in renderer
4. ✅ **Test with complex cases**: Added `FuncCall` with break propagation test

### Phase 2: AST Coverage Expansion (2-4 days)
1. **Add WHERE clause support**: `WhereClause`, `AExpr` ToTokens
2. **Add basic expressions**: `AConst`, binary operators
3. **Add INSERT/UPDATE/DELETE**: Basic DML statements

### Phase 3: Advanced Features (1-2 days)
1. **Implement group ID system**: Cross-group references
2. **Add performance optimizations**: Early bailout, caching
3. **Enhanced line breaking**: Better soft line semantics

### Phase 4: Production Ready (1-2 days)
1. **Comprehensive testing**: Large SQL files, edge cases
2. **Performance validation**: Benchmark against alternatives
3. **Documentation**: API docs, integration examples

## API Integration Points

```rust
// Main formatting function
pub fn format_sql(sql: &str, config: RenderConfig) -> Result<String, Error> {
    let parsed = pgt_query::parse(sql)?;
    let ast = parsed.root()?;
    
    let mut emitter = EventEmitter::new();
    ast.to_tokens(&mut emitter);
    
    let mut output = String::new();
    let mut renderer = Renderer::new(&mut output, config);
    renderer.render(emitter.events)?;
    
    Ok(output)
}

// Configuration
pub struct RenderConfig {
    pub max_line_length: usize,    // 80, 100, 120, etc.
    pub indent_size: usize,        // 2, 4, etc.
    pub indent_style: IndentStyle, // Spaces, Tabs
}
```

## Estimated Completion Timeline

- ✅ **Phase 1** (Core fixes): COMPLETED → **Fully functional renderer**
- **Phase 2** (AST coverage): 4 days → **Supports most common SQL**
- **Phase 3** (Advanced): 2 days → **Production-grade formatting**
- **Phase 4** (Polish): 2 days → **Integration ready**

**Total: ~1 week remaining** for complete production-ready PostgreSQL pretty printer.

## Current Limitations

1. **Limited SQL coverage**: Only basic SELECT statements and function calls
2. **No error recovery**: Unimplemented AST nodes cause panics  
3. **No configuration validation**: Invalid configs not checked
4. **Missing group ID system**: Cross-group conditional formatting not yet implemented

The core renderer foundation is now solid with proper break propagation and nested group handling - the remaining work is primarily expanding AST node coverage.