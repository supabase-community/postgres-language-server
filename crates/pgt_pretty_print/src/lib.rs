mod codegen;

pub use crate::codegen::token_kind::TokenKind;

use std::collections::VecDeque;

#[derive(Debug, Clone, PartialEq)]
pub enum LineType {
    /// Must break (semicolon, etc.)
    Hard,
    /// Break if group doesn't fit
    Soft,
    /// Break if group doesn't fit, but collapse to space if it does
    SoftOrSpace,
}

#[derive(Debug, Clone, PartialEq)]
pub enum LayoutEvent {
    Token(TokenKind),
    Space,
    Line(LineType),
    GroupStart {
        id: Option<String>,
        break_parent: bool,
    },
    GroupEnd,
    IndentStart,
    IndentEnd,
}

pub struct EventEmitter {
    pub events: VecDeque<LayoutEvent>,
}

impl EventEmitter {
    pub fn new() -> Self {
        Self {
            events: VecDeque::new(),
        }
    }

    /// Helper methods for emitting events
    pub fn token(&mut self, token: TokenKind) {
        self.events.push_back(LayoutEvent::Token(token));
    }

    pub fn space(&mut self) {
        self.events.push_back(LayoutEvent::Space);
    }

    pub fn line(&mut self, line_type: LineType) {
        self.events.push_back(LayoutEvent::Line(line_type));
    }

    pub fn group_start(&mut self, id: Option<String>, break_parent: bool) {
        self.events
            .push_back(LayoutEvent::GroupStart { id, break_parent });
    }

    pub fn group_end(&mut self) {
        self.events.push_back(LayoutEvent::GroupEnd);
    }

    pub fn indent_start(&mut self) {
        self.events.push_back(LayoutEvent::IndentStart);
    }

    pub fn indent_end(&mut self) {
        self.events.push_back(LayoutEvent::IndentEnd);
    }
}
