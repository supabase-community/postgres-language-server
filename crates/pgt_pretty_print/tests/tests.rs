use camino::Utf8Path;
use dir_test::{Fixture, dir_test};
use insta::{assert_snapshot, with_settings};

use pgt_pretty_print::{
    emitter::{EventEmitter, ToTokens},
    renderer::{IndentStyle, RenderConfig, Renderer},
};

#[dir_test(
    dir: "$CARGO_MANIFEST_DIR/tests/data/",
    glob: "*.sql",
)]
fn test_formatter(fixture: Fixture<&str>) {
    let content = fixture.content();
    let absolute_fixture_path = Utf8Path::new(fixture.path());
    let input_file = absolute_fixture_path;
    let test_name = absolute_fixture_path
        .file_name()
        .and_then(|x| x.strip_suffix(".sql"))
        .unwrap();

    // extract line length from filename (e.g., "simple_select_80" -> 80)
    let max_line_length = test_name
        .split('_')
        .next_back()
        .and_then(|s| s.parse::<usize>().ok())
        .unwrap_or(80);

    let parsed = pgt_query::parse(content).expect("Failed to parse SQL");
    let mut ast = parsed.into_root().expect("No root node found");

    let mut emitter = EventEmitter::new();
    ast.to_tokens(&mut emitter);

    let mut output = String::new();
    let config = RenderConfig {
        max_line_length,
        indent_size: 2,
        indent_style: IndentStyle::Spaces,
    };
    let mut renderer = Renderer::new(&mut output, config);
    renderer.render(emitter.events).expect("Failed to render");

    let parsed_output = pgt_query::parse(&output).expect("Failed to parse SQL");
    let mut parsed_ast = parsed_output.into_root().expect("No root node found");

    // the location fields are now different in the two ASTs
    clear_location(&mut parsed_ast);
    clear_location(&mut ast);

    assert_eq!(ast, parsed_ast);

    with_settings!({
      omit_expression => true,
      input_file => input_file,
    }, {
      assert_snapshot!(test_name, output);
    });
}

#[dir_test(
    dir: "$CARGO_MANIFEST_DIR/tests/data/",
    glob: "*.sql",
)]
fn validate_test_data(fixture: Fixture<&str>) {
    let content = fixture.content();
    let absolute_fixture_path = Utf8Path::new(fixture.path());
    let test_name = absolute_fixture_path
        .file_name()
        .and_then(|x| x.strip_suffix(".sql"))
        .unwrap();

    let result = pgt_query::parse(content);

    if let Ok(res) = result.as_ref() {
        assert!(res.root().is_some(), "should have a single root node");
    }

    assert!(result.is_ok(), "should be valid SQL");
}

fn clear_location(node: &mut pgt_query::NodeEnum) {
    unsafe {
        node.iter_mut().for_each(|n| match n {
            pgt_query::NodeMut::ColumnRef(n) => {
                (*n).location = 0;
            }
            pgt_query::NodeMut::ParamRef(n) => {
                (*n).location = 0;
            }
            pgt_query::NodeMut::AExpr(n) => {
                (*n).location = 0;
            }
            pgt_query::NodeMut::TypeCast(n) => {
                (*n).location = 0;
            }
            pgt_query::NodeMut::CollateClause(n) => {
                (*n).location = 0;
            }
            pgt_query::NodeMut::FuncCall(n) => {
                (*n).location = 0;
            }
            pgt_query::NodeMut::AArrayExpr(n) => {
                (*n).location = 0;
            }
            pgt_query::NodeMut::ResTarget(n) => {
                (*n).location = 0;
            }
            pgt_query::NodeMut::SortBy(n) => {
                (*n).location = 0;
            }
            pgt_query::NodeMut::WindowDef(n) => {
                (*n).location = 0;
            }
            pgt_query::NodeMut::TypeName(n) => {
                (*n).location = 0;
            }
            pgt_query::NodeMut::ColumnDef(n) => {
                (*n).location = 0;
            }
            pgt_query::NodeMut::DefElem(n) => {
                (*n).location = 0;
            }
            pgt_query::NodeMut::XmlSerialize(n) => {
                (*n).location = 0;
            }
            pgt_query::NodeMut::AConst(n) => {
                (*n).location = 0;
            }
            pgt_query::NodeMut::RangeVar(n) => {
                (*n).location = 0;
            }
            _ => {}
        });
    }
}
