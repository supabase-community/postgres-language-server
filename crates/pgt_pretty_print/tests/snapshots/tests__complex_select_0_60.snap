---
source: crates/pgt_pretty_print/tests/tests.rs
input_file: crates/pgt_pretty_print/tests/data/complex_select_0_60.sql
snapshot_kind: text
---
WITH RECURSIVE pks_fks AS (SELECT
  contype::text AS contype,
  conname,
  array_length(
    conkey,
    1
  ) AS ncol,
  conrelid AS resorigtbl,
  col AS resorigcol,
  ord
FROM
  pg_constraint
  LEFT JOIN
    LATERAL unnest(
      conkey
    ) WITH ORDINALITY AS _(col, ord) ON TRUE
WHERE contype IN ('p', 'f')
UNION
SELECT
  concat(
    contype,
    '_ref'
  ) AS contype,
  conname,
  array_length(
    confkey,
    1
  ) AS ncol,
  confrelid,
  col,
  ord
FROM
  pg_constraint
  LEFT JOIN
    LATERAL unnest(
      confkey
    ) WITH ORDINALITY AS _(col, ord) ON TRUE
WHERE contype = 'f'), views AS (SELECT
  c.oid AS view_id,
  n.nspname AS view_schema,
  c.relname AS view_name,
  r.ev_action AS view_definition
FROM
  pg_class c
  INNER JOIN
    pg_namespace n ON n.oid = c.relnamespace
  INNER JOIN
    pg_rewrite r ON r.ev_class = c.oid
WHERE c.relkind IN ('v', 'm')), transform_json AS (SELECT
  view_id,
  view_schema,
  view_name,
  replace(
    replace(
      replace(
        replace(
          replace(
            replace(
              replace(
                regexp_replace(
                  replace(
                    replace(
                      replace(
                        replace(
                          replace(
                            replace(
                              replace(
                                replace(
                                  replace(
                                    replace(
                                      replace(
                                        view_definition::text,
                                        '<>',
                                        '()'
                                      ),
                                      ',',
                                      ''
                                    ),
                                    '\\{',
                                    ''
                                  ),
                                  '\\}',
                                  ''
                                ),
                                ' :targetList ',
                                ',"targetList":'
                              ),
                              ' :resno ',
                              ',"resno":'
                            ),
                            ' :resorigtbl ',
                            ',"resorigtbl":'
                          ),
                          ' :resorigcol ',
                          ',"resorigcol":'
                        ),
                        '{',
                        '{ :'
                      ),
                      '((',
                      '{(('
                    ),
                    '({',
                    '{({'
                  ),
                  ' :[^}{,]+',
                  ',"":',
                  'g'
                ),
                ',"":}',
                '}'
              ),
              ',"":,',
              ','
            ),
            '{(',
            '('
          ),
          '{,',
          '{'
        ),
        '(',
        '['
      ),
      ')',
      ']'
    ),
    ' ',
    ','
  )::pg_catalog.json AS view_definition
FROM
  views), target_entries AS (SELECT
  view_id,
  view_schema,
  view_name,
  json_array_elements(
    view_definition -> 0 -> 'targetList'
  ) AS entry
FROM
  transform_json), results AS (SELECT
  view_id,
  view_schema,
  view_name,
  (entry ->> 'resno')::pg_catalog.int4 AS view_column,
  (entry ->> 'resorigtbl')::oid AS resorigtbl,
  (entry ->> 'resorigcol')::pg_catalog.int4 AS resorigcol
FROM
  target_entries), recursion(view_id, view_schema, view_name, view_column, resorigtbl, resorigcol, is_cycle, path) AS (SELECT
  r.*,
  FALSE,
  ARRAY[resorigtbl]
FROM
  results r
WHERE TRUE
UNION ALL
SELECT
  view.view_id,
  view.view_schema,
  view.view_name,
  view.view_column,
  tab.resorigtbl,
  tab.resorigcol,
  tab.resorigtbl = ANY(path),
  path || tab.resorigtbl
FROM
  recursion view
  INNER JOIN
    results tab ON view.resorigtbl = tab.view_id
      AND view.resorigcol = tab.view_column
WHERE NOT is_cycle), repeated_references AS (SELECT
  view_id,
  view_schema,
  view_name,
  resorigtbl,
  resorigcol,
  array_agg(
    attname
  ) AS view_columns
FROM
  recursion
  INNER JOIN
    pg_attribute vcol ON vcol.attrelid = view_id
      AND vcol.attnum = view_column
GROUP BY view_id, view_schema, view_name, resorigtbl, resorigcol)
SELECT
  sch.nspname AS table_schema,
  tbl.relname AS table_name,
  rep.view_schema,
  rep.view_name,
  pks_fks.conname AS constraint_name,
  pks_fks.contype AS constraint_type,
  jsonb_agg(
    jsonb_build_object(
      'table_column',
      col.attname,
      'view_columns',
      view_columns
    )
  ORDER BY pks_fks.ord) AS column_dependencies
FROM
  repeated_references rep
  INNER JOIN
    pks_fks USING (resorigtbl, resorigcol)
  INNER JOIN
    pg_class tbl ON tbl.oid = rep.resorigtbl
  INNER JOIN
    pg_attribute col ON col.attrelid = tbl.oid
      AND col.attnum = rep.resorigcol
  INNER JOIN
    pg_namespace sch ON sch.oid = tbl.relnamespace
GROUP BY sch.nspname, tbl.relname, rep.view_schema, rep.view_name, pks_fks.conname, pks_fks.contype, pks_fks.ncol
HAVING ncol = array_length(
  array_agg(
    ROW(col.attname, view_columns)
  ORDER BY pks_fks.ord),
  1
);
